### PR TITLE
Deprecate "local" extension types

### LINKED ISSUE #20540: Deprecate extensions folder structure
In order to make programmatic management of extensions more stable, we gotta:

- Drop the individual type packages in extensions (eg interfaces/displays/etc)
- Require `package.json` to exist for every extension
- Drop the requirement for root extensions to start with `directus-extension-*`

```[tasklist]
### Documentation
- [ ] https://github.com/directus/directus/issues/20622
- [ ] https://github.com/directus/directus/issues/20623
```

### PR DESCRIPTION
**Pull request by rijkvanzanten**
## Scope

What's changed:

- Drop support for the outdated "local" per-folder extension type

## Potential Risks / Drawbacks

- It's a breaking change for people with existing local folder-style extensions. This should not be shipped to production without the proper migration guide and announcements in place (@phazonoverload owns that)

## Review Notes / Questions

- If you're concerned about the above listed risk, this is merging to `next` rather than `main` for that exact reason 👍🏻 
- I have _not_ properly kicked the tires on this yet. I just removed the chunks of code I thought where related to the local extension type stuff. I'd love a quick look from @nickrum and @licitdev to make sure I didn't miss anything, and then we can all QA the actual usability of extensions afterwards 🙂 

---

Fixes #20540

### CODE REVIEW

**Review by nickrum — DISMISSED** [review]
Do we want to keep the `directus-extension-` prefix for npm extensions so we can distinguish regular packages from actual Directus extensions and easily search for them? After all, that's why it was added in the first place.

**Comment by paescuj** [comment]
> Do we want to keep the `directus-extension-` prefix for npm extensions so we can distinguish regular packages from actual Directus extensions and easily search for them? After all, that's why it was added in the first place.

Yes 👍

**Comment by rijkvanzanten** [comment]
See https://github.com/directus/directus/commit/0f3f926105fdb3e86f1b82dc2dc582f5d2653e61 👆🏻

**Comment by nickrum** [comment]
How would you find them based on package.json contents in the npm registry?

**Comment by paescuj** [comment]
> I wanna do it by keyword yeah. That's how you "tag" things on npm, making it feel very appropriate for this use case

Good to know 😀 I was particularly concerned about how we can do the lookup in the registry, but if the plan is to do that by keyword, I guess that works.

**Comment by NeoKactus** [comment]
Will this MR account for the extension auto-reload issue `directus-extension-*` had? https://github.com/directus/directus/issues/20380

It was one of the reasons we stuck with the "local" per-folder extension type pattern.

**Comment by paescuj** [comment]
> Will this MR account for the extension auto-reload issue `directus-extension-*` had? #20380

I don't think so, however this issue is something we will look at 👍

**Comment by rijkvanzanten** [comment]
@nickrum Reverted the dependency based logic to use the name prefix 👍🏻

**Comment by rijkvanzanten** [comment]
@nickrum / @paescuj This should be ready to merge now 🚀

File: api/src/extensions/lib/get-extensions.ts [hunk 1]
```diff
@@ -1,54 +1,25 @@
...
-import type { Extension } from '@directus/extensions';
-import { getLocalExtensions, getPackageExtensions, resolvePackageExtensions } from '@directus/extensions/node';
+import { resolveDependencyExtensions, resolveExtensions } from '@directus/extensions/node';
 import { useLogger } from '../../logger.js';
 import { getExtensionsPath } from './get-extensions-path.js';
 
 export const getExtensions = async () => {
 	const env = useEnv();
 	const logger = useLogger();
+	const extensions = await resolveExtensions(getExtensionsPath());
```
Comment (paescuj) [inline api/src/extensions/lib/get-extensions.ts:9]: Would prefer if we use the term "local extensions" to make it clearer and also consistent with `vite.config.js`. 
```suggestion
	const localExtensions = await resolveLocalExtensions(getExtensionsPath());
```

File: api/src/extensions/lib/get-extensions.ts [hunk 2]
```diff
@@ -1,54 +1,25 @@
...
-					// Do not load entire bundle if it has duplicated entries
-					duplicateExtensions.push(extension.name);
-					return;
-				}
-
-				bundleEntryNames.add(entry.name);
-			}
+	const dedupedDependencyExtensions = dependencyExtensions.filter((ext) => {
+		if (extensionNames.includes(ext.name)) {
+			logger.warn(`Node module extension "${ext.name}" already exists within the extensions folder.`);
```
Comment (paescuj) [inline api/src/extensions/lib/get-extensions.ts:17]: Not sure, but would probably speak of a package extension
```suggestion
			logger.warn(`Package extension "${ext.name}" already exists within the local extensions folder.`);
```

File: app/vite.config.js [hunk 3]
```diff
@@ -20,8 +13,16 @@ import { searchForWorkspaceRoot } from 'vite';
 import { defineConfig } from 'vitest/config';
 
 const API_PATH = path.join('..', 'api');
+
+/*
+ * @TODO This extension path is hardcoded to the env default (./extensions). This won't work
+ * as expected when extensions are read from a different location locally through the
+ * EXTENSIONS_LOCATION env var
+ */
```
Comment (paescuj) [inline app/vite.config.js:21]: Couldn't we use the new `@directus/env` package here? (👀 🚀)

File: app/vite.config.js [hunk 4]
```diff
@@ -136,12 +137,16 @@ function directusExtensions() {
...
+		const dependencyExtensions = await resolveDependencyExtensions(API_PATH);
 
-		const extensions = [...packageExtensions, ...localPackageExtensions, ...localExtensions];
+		/*
+		 * @TODO
+		 * These aren't deduplicated, whereas the production ones are. This has seemingly
+		 * always been the case. Is this a bug?
+		 * @see /api/src/extensions/lib/get-extensions.ts
+		 */
+		const extensions = [...localExtensions, ...dependencyExtensions];
```
Comment (paescuj) [inline app/vite.config.js:149]: Would it make sense / be possible to share the code from `get-extensions.ts`?

File: api/src/extensions/lib/get-extensions.ts (hunk 1 above)
Comment (rijkvanzanten) [reply api/src/extensions/lib/get-extensions.ts:9]: Deal 🤝

**Comment by paescuj** [comment]
Leaving as a note for the migration docs:

The typed extensions folders must be completely removed! Otherwise the server won't start:
```
Error: Extension "displays" does not contain a package.json file.
```

File: packages/extensions/src/node/utils/get-extensions.ts [hunk 5]
```diff
@@ -98,58 +99,23 @@ export async function resolvePackageExtensions(root: string, extensionNames?: st
 	return extensions;
 }
 
-export async function getPackageExtensions(root: string): Promise<Extension[]> {
+const isExtension = (pkgName: string) => {
+	const regex = /^(?:(?:@[^/]+\/)?directus-extension-|@directus\/extension-)(.+)$/;
```
Comment (DanielBiegler) [inline packages/extensions/src/node/utils/get-extensions.ts:103]: This is probably not important so feel free to ignore but I think the following might be "more correct" as this disallows multiple leading @ symbols:

```suggestion
	const regex = /^(?:(?:@[^/@]+\/)?directus-extension-|@directus\/extension-)(.+)$/;
```

Comment (paescuj) [reply packages/extensions/src/node/utils/get-extensions.ts:101]: Technically, all non-URL-safe characters are disallowed, but since dependency names are ensured to be valid via package manager / NPM we can be more lax here 👍 

```console
❯ grep name packages/test/package.json 
  "name": "te@st",

❯ pn add -w te@st
 ERR_PNPM_NO_MATCHING_VERSION  No matching version found for te@st
```
(^ although the package "exists" locally)

File: packages/extensions-sdk/src/cli/commands/add.ts [hunk 6]
```diff
@@ -175,8 +174,6 @@ export default async function add(options: AddOptions): Promise<void> {
 			process.exit(1);
 		}
 
-		const oldName = extensionManifest.name.match(EXTENSION_NAME_REGEX)?.[1] ?? extensionManifest.name;
-
```
Comment (paescuj) [inline packages/extensions-sdk/src/cli/commands/add.ts:179]: I think it would make sense to revert this change, following 35714f6a4efe67d071df3de9d6bc8a814bded5dd

File: packages/extensions-sdk/src/cli/commands/create.ts [hunk 7]
```diff
@@ -177,7 +176,7 @@ async function createLocalExtension({
 
 function getPackageManifest(name: string, options: ExtensionOptions, deps: Record<string, string>) {
 	const packageManifest: Record<string, any> = {
-		name: EXTENSION_NAME_REGEX.test(name) ? name : `directus-extension-${name}`,
```
Comment (paescuj) [inline packages/extensions-sdk/src/cli/commands/create.ts:180]: Same as in `packages/extensions-sdk/src/cli/commands/add.ts`:

> I think it would make sense to revert this change, following 35714f6a4efe67d071df3de9d6bc8a814bded5dd

**Review by paescuj — APPROVED** [review]
Non-blocking change request to partially revert changes in `extensions-sdk`.
The pending comments in `@directus/app` are also non-blocking.
Which means: LGTM 🚀 

@phazonoverload
Do we already have an issue for the docs implication?
I want to make sure https://github.com/directus/directus/pull/20624#issuecomment-1882133888 will not be forgotten.

File: packages/extensions-sdk/src/cli/commands/add.ts (hunk 6 above)
Comment (rijkvanzanten) [reply packages/extensions-sdk/src/cli/commands/add.ts:179]: I'm not sure if that's relevant anymore, as it's only the package extensions that are checking the name now 🤔 This is only for adding entries to a bundle IIRC

File: app/vite.config.js (hunk 4 above)
Comment (rijkvanzanten) [reply app/vite.config.js]: No, that's using the APIs logger as well (although.... refactoring that to @directus/logger would be nice)

File: app/vite.config.js (hunk 3 above)
Comment (rijkvanzanten) [reply app/vite.config.js:21]: I don't know if that matters 🤔 This is really only meant for dev mode when running Directus from the repo. Ideally this is removed completely from vite and instead handled through the API just like in production mode!

Comment (paescuj) [reply app/vite.config.js:21]: Yup 👍
